logs_sdk/lib/opentelemetry/sdk/logs/logger_provider.rb

```diff
@@ -0,0 +1,129 @@
+
+        EMPTY_NAME_ERROR = 'LoggerProvider#logger called without '\
+          'providing a logger name.'
+        FORCE_FLUSH_ERROR = 'unexpected error in ' \
+          'OpenTelemetry::SDK::Logs::LoggerProvider#force_flush'
+
+        # Returns a new LoggerProvider instance.
+        #
+        # @param [optional Resource] resource The resource to associate with
+        #   new LogRecords created by {Logger}s created by this LoggerProvider.
+        # @param [optional Array] log_record_processors The
+        #   {LogRecordProcessor}s to associate with this LoggerProvider.
+        #
+        # @return [OpenTelemetry::SDK::Logs::LoggerProvider]
+        def initialize(
+          resource: OpenTelemetry::SDK::Resources::Resource.create,
+          log_record_processors: []
+        )
+          @log_record_processors = log_record_processors
+          @mutex = Mutex.new
+          @resource = resource
+          @stopped = false
+        end
+
+        # Creates an {OpenTelemetry::SDK::Logs::Logger} instance.
+        #
+        # @param [optional String] name Instrumentation package name
+        # @param [optional String] version Instrumentation package version
+        #
+        # @return [OpenTelemetry::SDK::Logs::Logger]
+        def logger(name = nil, version = nil)
+          name ||= ''
+          version ||= ''
+
+          OpenTelemetry.logger.warn(EMPTY_NAME_ERROR) if name.empty?
+
+          @mutex.synchronize do
+            OpenTelemetry::SDK::Logs::Logger.new(name, version, self)
+          end
+        end
+
+        # Adds a new log record processor to this LoggerProvider's
+        # log_record_processors.
+        #
+        # @param [LogRecordProcessor] log_record_processor The
+        #   {LogRecordProcessor} to add to this LoggerProvider.
+        def add_log_record_processor(log_record_processor)
+          @mutex.synchronize do
+            @log_record_processors = log_record_processors.dup.push(log_record_processor)
+          end
+        end
+
+        # Attempts to stop all the activity for this LoggerProvider. Calls
+        # {LogRecordProcessor#shutdown} for all registered {LogRecordProcessor}s.
+        #
+        # This operation may block until all log records are processed. Must
+        # be called before turning off the main application to ensure all data
+        # are processed and exported.
+        #
+        # After this is called all newly created {LogRecord}s will be no-op.
+        #
+        # @param [optional Numeric] timeout An optional timeout in seconds.
+        # @return [Integer] Export::SUCCESS if no error occurred, Export::FAILURE if
+        #   a non-specific failure occurred, Export::TIMEOUT if a timeout occurred.
+        def shutdown(timeout: nil)
+          @mutex.synchronize do
+            if @stopped
+              OpenTelemetry.logger.warn('LoggerProvider#shutdown called multiple times.')
+              return OpenTelemetry::SDK::Logs::Export::FAILURE
```

Comment:
The code for `LoggerProvider#shutdown` and `LoggerProvider#force_flush` are drawn mainly from `TracerProvider` methods of the same names. 

Why does `#shutdown` return a failure status code if called when `@stopped` is true, but `#force_flush` returns a success status code?

JS, [for example](https://github.com/open-telemetry/opentelemetry-js/blob/b400c2e5d9729c3528482781a93393602dc6dc9f/experimental/packages/sdk-logs/src/LoggerProvider.ts#L124-L145 ), does similar things for both `#force_flush` and `#shutdown`.

The specs on `#force_flush` (very similar for [Logs](https://opentelemetry.io/docs/specs/otel/logs/sdk/#forceflush) and [Traces](https://opentelemetry.io/docs/specs/otel/trace/sdk/#forceflush)), don't describe the desired behavior in cases where the method is called, but the provider has stopped/shutdown.

Reply:
I think there's some flexibility in the spec language around the behavior of shutdown. It says:

> After the call to Shutdown, subsequent attempts to get a Logger are not allowed. 

In java, we log an error message if when shutdown is called multiple times, but don't return an error. I think the intent of "not allowed" is to avoid performing the behavioral side effects of calling shutdown. Specifically, called shutdown on the provider is support to be implemented by calling shutdown on all registered processors, which we don't want to do multiple times.

Reply:
Thanks for clarifying! It sounds like Ruby is in alignment as-is.